accounts/tests.py

```diff
@@ -1,15 +1,229 @@
 import datetime
+from unittest.mock import Mock, patch
 
+from django.contrib.flatpages.models import FlatPage
+from django.contrib.sessions.backends.db import SessionStore
+from django.contrib.sites.models import Site
+from django.http import HttpRequest
 from django.test import TestCase
+from django.urls import reverse
 from django_dynamic_fixture import G
 from lark.exceptions import UnexpectedCharacters
 
-from accounts.models import Child, User
+from accounts.backends import TWO_FACTOR_AUTH_SESSION_KEY
+from accounts.models import Child, GoogleAuthenticatorTOTP, User
 from accounts.queries import get_child_eligibility
 from studies.fields import GESTATIONAL_AGE_CHOICES
 from studies.models import Lab, Study, StudyType
 
 
+class AuthenticationTestCase(TestCase):
+    def setUp(self):
+        self.researcher_email = "test@test.com"
+        self.participant_email = "tom@myspace.com"
+        self.test_password = "testpassword20chars"
+        self.base32_secret = "GNKVX3Y2U6BKTVKU"
+
+        # Researcher setup
+        self.researcher = G(
+            User,
+            username=self.researcher_email,
+            is_active=True,
+            is_researcher=True,
+            nickname="Lab Researcher",
+        )
+        self.researcher.set_password(self.test_password)
+        self.otp = GoogleAuthenticatorTOTP.objects.create(
+            user=self.researcher, secret=self.base32_secret, activated=True
+        )
+        self.researcher.save()
+
+        # Participant Setup
+        self.participant = G(
+            User,
+            username=self.participant_email,
+            is_active=True,
+            is_researcher=False,
+            nickname="MySpace Tom",
+        )
+        self.participant.set_password(self.test_password)
+        self.participant.save()
+
+        # Site fixture enabling login
+        self.fake_site = G(Site, id=1)
+
+        # FlatPage fixture enabling login redirect to work.
+        self.home_page = G(FlatPage, url="/")
+        self.home_page.sites.add(self.fake_site)
+        self.home_page.save()
+
+    def test_researcher_registration_flow(self):
+        response = self.client.post(
+            reverse("accounts:researcher-registration"),
+            {
+                "username": "tester@test.com",
+                "password1": self.test_password,
+                "password2": self.test_password,
+                "given_name": "Test",
+                "family_name": "Person",
+                "nickname": "Testman",
+            },
+            follow=True,
+        )
+
+        self.assertEqual(
+            response.redirect_chain, [(reverse("accounts:2fa-setup"), 302)]
+        )
+        self.assertEqual(response.status_code, 200)
+
+        # Mock correctly entered OTP
+        otp = response.context["user"].otp
+        response = self.client.post(
+            reverse("accounts:2fa-setup"), {"otp_code": otp.provider.now()}, follow=True
```

Comment:
I know I'm likely impressed by the wrong things but it's very cool that we're able to test this out.